internals-js/src/extractSubgraphsFromSupergraph.ts

```diff
@@ -625,51 +577,360 @@ function errorToString(e: any,): string {
+   */
+  const reachableTypesBySubgraph = collectFieldReachableTypesForAllSubgraphs(
+    supergraph,
+    subgraphs.names(),
+    (f, name) => {
+      const fieldApplications: Directive<any, { graph?: string, requires?: string, provides?: string }>[] = f.appliedDirectivesOf(fieldDirective);
+      if (fieldApplications.length) {
+        const application = fieldApplications.find((application) => getSubgraph(application)?.name === name);
+        if (application) {
+          const args = application.arguments();
+          const typesInFederationDirectives =
+            typesUsedInFederationDirective(args.provides, baseType(f.type!) as CompositeType)
+            .concat(typesUsedInFederationDirective(args.requires, f.parent));
+          return { isInSubgraph: true, typesInFederationDirectives };
+        } else {
+          return { isInSubgraph: false, typesInFederationDirectives: [] };
+        }
+      } else {
+        // No field application depends on the "owner" directive on the type. If we have no owner, then the
+        // field is in all subgraph and we return true. Otherwise, the field is only in the owner subgraph.
+        // In any case, the field cannot have a requires or provides
+        const ownerApplications = ownerDirective ? f.parent.appliedDirectivesOf(ownerDirective) : [];
+        return { isInSubgraph: !ownerApplications.length || getSubgraph(ownerApplications[0])?.name == name, typesInFederationDirectives: [] };
+      }
+    },
+    (t, name) => {
+      const typeApplications: Directive<any, { graph: string, key?: string}>[] = t.appliedDirectivesOf(typeDirective);
+      const application = typeApplications.find((application) => (application.arguments().key && (getSubgraph(application)?.name === name)));
+      if (application) {
+        const typesInFederationDirectives = typesUsedInFederationDirective(application.arguments().key, t as CompositeType);
+        return { isEntityWithKeyInSubgraph: true, typesInFederationDirectives };
+      } else {
+        return { isEntityWithKeyInSubgraph: false, typesInFederationDirectives: [] };
+      }
+    },
+  );
+  const includeTypeInSubgraph = (t: NamedType, name: string) => reachableTypesBySubgraph.get(name)?.has(t.name) ?? false;
+
+  // Next, we iterate on all types and add it to the proper subgraphs (along with any @key).
+  // Note that we first add all types empty and populate the types next. This avoids having to care about the iteration
+  // order if we have fields than depends on other types.
+  for (const type of filteredTypes) {
+    const typeApplications = type.appliedDirectivesOf(typeDirective);
+    if (!typeApplications.length) {
+      // Imply we don't know in which subgraph the type is, so we had it in all subgraph in which the type is reachable.
+      subgraphs
+        .values()
+        .filter((sg) => includeTypeInSubgraph(type, sg.name))
+        .map(sg => sg.schema).forEach(schema => schema.addType(newNamedType(type.kind, type.name)));
```

Comment:
I know this is previously existing code but this filter/map is pretty trivial and adds 100s of iterations (in the case we're hearing about lately). Fine to leave as is but this looks like an easy win.
```suggestion
        .filter((sg) => includeTypeInSubgraph(type, sg.name))
        .map(sg => sg.schema)
        
        .forEach(sg => {
          if (!includeTypeInSubgraph(type, sg.name)) return;
          sg.schema.addType(newNamedType(type.kind, type.name));
        });
```

Reply:
I'm ok changing that. Made it a good'ol for-loop while at it though as it's imo more readable at that point.

I'll note though that this is code that only run for fed1 supergraphs, so only for use that never ever composed with the fed2 composition code, and so not necessary where performance matters much.